### ISSUE TITLE
Multi-argument `eachindex`

### ISSUE THREAD

**Issue by lindahua** [JuliaLang/julia#10906]
The `eachindex` is a great function, which provides a uniform & efficient way to traverse a array. 

However, in many cases, the computation require multiple arguments, _e.g._, addition. 

``` julia
s = rand(5, 6)
a = rand(4, 6)   # with fast linear indexing
b = sub(s, 1:4, :)   # no fast linear indexing
c = zeros(4, 6)

# It would be great to do the following
for (ia, ib, ic) in eachindex(a, b, c)
    c[ic] = a[ia] + b[ib]
end
```

A more greedy demand is to allow broadcasting eachindex when arrays have different sizes, but their shapes are compatible (in terms of broadcasting).

cc: @timholy

**Comment by timholy** [comment]
Agreed on both points. The two-argument version was just implemented in #10858, but I agree we could probably use a multi-argument version. For the broadcasting idea, there's something like that [in reducedim](https://github.com/JuliaLang/julia/blob/438e90bb2563d653c0fa6ebee433e3e83836fb60/base/reducedim.jl#L198-L216), but I think it only adds trailing 1s. You can use `max(I1,I2)` on two `CartesianIndexes`. I bet all this is not as flexible or easy to use as it should be, however.

**Comment by tkelman** [comment]
doesn't seem release-blocking
